### ISSUE TITLE
Observers/reactives fire in indeterminate order

### ISSUE THREAD

**Issue by wch** [rstudio/shiny#2466]
This is a follow-up to #2463 and https://github.com/rstudio/shiny/pull/2429#issuecomment-497474108.

We've always told users that shouldn't rely on the reactives/observers firing in a specific order (unless `priority` is used), because the order is arbitrary. However, it's also true that we've assumed that the observers will fire in a deterministic order -- specifically, the order in which they were defined.

This assumptions turns out to not be true, either in the CRAN version of Shiny (1.3.2) or the current master version of shiny, with the fastmap PR merged (#2429).

Here's an example:

```R
library(shiny)

v <- reactiveVal(0)

observe({
  message(v(), ": observer 1")
})
observe({
  message(v(), ": observer 2")
})
observe({
  message(v(), ": observer 3")
})


v(isolate(v()) + 1); shiny:::flushReact()
v(isolate(v()) + 1); shiny:::flushReact()
v(isolate(v()) + 1); shiny:::flushReact()
# and so on ...
```

The naive assumption is that it would print out something like this each time `flushReact()` is called:

```
> v(isolate(v()) + 1); shiny:::flushReact()
1: observer 1
1: observer 2
1: observer 3
> v(isolate(v()) + 1); shiny:::flushReact()
2: observer 1
2: observer 2
2: observer 3
```

And in the CRAN version of Shiny, it does this... for a while. But after running it a number times, it changes order:

```
d> v(isolate(v()) + 1); shiny:::flushReact()
6: observer 3
6: observer 1
6: observer 2

...

d> v(isolate(v()) + 1); shiny:::flushReact()
19: observer 2
19: observer 3
19: observer 1
```

In the current master, it does something different:

```
> v(isolate(v()) + 1); shiny:::flushReact()
1: observer 1
1: observer 2
1: observer 3
> v(isolate(v()) + 1); shiny:::flushReact()
2: observer 3
2: observer 2
2: observer 1
> v(isolate(v()) + 1); shiny:::flushReact()
3: observer 1
3: observer 2
3: observer 3
> v(isolate(v()) + 1); shiny:::flushReact()
4: observer 3
4: observer 2
4: observer 1
```

It switches back and forth, reversing the order each time.

The question at this point is what to do about it. We can make the order fixed, by changing this line:

https://github.com/rstudio/shiny/blob/17afce6fa1f3bc284a295b2ed16fb143692fbc79/R/reactives.R#L47

from this:

```
  .dependents$values(),
```

to something like this:

```
   .dependents$mget(as.character(sort(as.numeric(.dependents$keys())))),
```

With that change, it goes in 1-2-3 order every time.

However, another option is to just leave it as is on master. With the CRAN version of Shiny, it looks like applications will fire the observers in the defined order many times, and then switch. For app developers, if their app relies on the ordering assumption, they might not test their application enough to see that it falls down, or that when it does, it happens inconsistently enough that they don't know how to debug it.

With the master version of Shiny, the order switches every time, so if a developer relies on ordering, it's a fail-fast result, which can help them find the bug more quickly.

So in short: we can sort the dependents, in which case observers fire in the order that they're invalidated (I think), and this will allow developers to rely on at some (undocumented) assumptions about ordering. Or we can not sort them, in which case developers will quickly encounter problems if they rely on undocumented assumptions about ordering, and have to fix their code.

Keep in mind that for any existing apps that rely on the ordering assumption, with the CRAN version of Shiny, they are currently buggy, even if it is not immediately obvious to the developer.

**Comment by wch** [comment]
One more note: on master we currently have tests that work on my Mac but fail on Travis, because the tests assume a certain order -- that order is correct on my Mac, but apparently not on Travis. See https://travis-ci.org/rstudio/shiny/jobs/539878522#L1254-L1270

The order of items and keys from a fastmap is dependent on the platform's implementation of `std::unordered_map`, and it apparently is different between my Mac and the Linux machines used by Travis.

**Comment by wch** [comment]
Now that I think about it some more, it seems problematic that the behavior is platform-dependent. That means that it might behave one way on the development platform, and another way on the deployment platform. I think there were even some unit tests that passed on macOS 10.13, but failed after I installed 10.14 and rebuilt the package.

With this in mind, I think it makes sense to either add the sorting to the `Dependents` class as described above, or possibly in the `Map` class, or even in fastmap itself.
